ArloCameraSource.js

```diff
@@ -0,0 +1,569 @@
+        this.lastSnapshot = null;
+
+        this.videoProcessor = config.videoProcessor || pathToFfmpeg || 'ffmpeg';
+        this.videoDecoder = config.videoDecoder || '';
+        this.videoEncoder = config.videoEncoder || 'libx264';
+        this.audioCodec = config.audioEncoder || 'libopus';
+        this.packetsize = config.packetsize || 1316; //188, 376, 1316
+        this.fps = 24;
+        this.maxBitrate = config.maxBitrate || 300;
+        this.additionalVideoCommands = (config.additionalVideoCommands ? (' ' + config.additionalVideoCommands) : '');
+        this.additionalAudioCommands = (config.additionalAudioCommands ? (' ' + config.additionalAudioCommands) : '');
+
+        let numberOfStreams = config.maxStreams || 2;
+
+        let options = {
+            proxy: false, // Requires RTP/RTCP MUX Proxy
+            srtp: true, // Supports SRTP AES_CM_128_HMAC_SHA1_80 encryption
+            video: {
+                resolutions: [
+                    [1280, 720, 24],
+                    [1280, 720, 15],
+                    [640, 360, 24],
+                    [640, 360, 15],
+                    [320, 240, 24],
+                    [320, 240, 15]
+                ],
+                codec: {
+                    profiles: [0, 1, 2],    //[StreamController.VideoCodecParamProfileIDTypes.MAIN],
+                    levels: [0, 1, 2]       //[StreamController.VideoCodecParamLevelTypes.TYPE4_0]
+                }
+            },
+            audio: {
+                codecs: [
+                    {
+                        type: 'OPUS',
+                        samplerate: 24,
+                        comfort_noise: true
+                    }
+                ]
+            }
+        }
+
+        this._createStreamControllers(numberOfStreams, options);
+        debug('Generated Camera Controller');
+    }
+
+    handleCloseConnection(connectionID) {
+        this.streamControllers.forEach(function(controller) {
+            controller.handleCloseConnection(connectionID);
+        });
+    }
+
+    handleSnapshotRequest(request, callback) {
+        debug('Snapshot requested');
+        let now = Date.now();
+
+        if (this.lastSnapshot && now < this.lastSnapshot + 300000) {
+            this.log('Snapshot skipped: Camera %s [%s] - Next in %d secs', this.accessory.displayName, this.device.id, parseInt((this.lastSnapshot + 300000 - now) / 1000));
+            callback();
+            return;
+        }
+
+        this.log("Snapshot request: Camera %s [%s]", this.accessory.displayName, this.device.id);
+
+        this.device.getSnapshot(function(error, data) {
+            if (error) {
+                this.log(error);
+                callback();
+                return;
+            }
+
+            this.lastSnapshot = Date.now();
+
+            this.log("Snapshot confirmed: Camera %s [%s]", this.accessory.displayName, this.device.id);
+
+           this.device.once(Arlo.FF_SNAPSHOT, function(url) {
```

Comment:
Calling getSnapshot or once(Arlo.FF_SNAPSHOT), not sure which exactly, will trigger a camera shutter on my first gen Arlo cam. It drains battery and somehow streaming stops working after a few minutes of restarting homebridge. We had this issue in homebridge-arlo v0.0.3 straeming-pr branch.

In my fork I just call downloadSnapshot(this.device.device.presignedLastImageUrl) without triggering a camera shutter. The snapshot is whatever available on Arlo server, but streaming works really well. I can connect to the camera any time in Home app. In case this is what's blocking v3.0.0, I've included the link to my change.

https://github.com/cnnn/homebridge-arlo/commit/bc163b1fb40b4e721be22e7166b520acc440842b#diff-2c2fd0251ce73d5331dc30d63853ba6dR286

Btw Thank You for maintaining the plugin!

Reply:
As you said, snapshots are a lot better, but streaming doesn't work.